### ISSUE TITLE
[bug] Spacebattles and SufficientVelocity broke with 'href' attribute error

### ISSUE THREAD

**Issue by tr13dm03** [JimmXinu/FanFicFare#207]
Found it is now impossible to update or download new fiction from threads on SV and SB. QQ is unaffected.

Relevant logs below. The threads used for testing simply happened to be in the front page and have proper indexes.

``` log
Unable to open ~/.mtpz-data for reading, MTPZ disabled.
FFF: DEBUG: 2017-08-03 01:47:31,422: calibre_plugins.fanficfare_plugin.fff_plugin(207): Plugin FanFicFare macmenuhack file_path:/home/user/.config/calibre/plugins/fanficfare_macmenuhack.txt
FFF: DEBUG: 2017-08-03 01:47:37,513: calibre_plugins.fanficfare_plugin.fff_plugin(954): FanFicFare v2.15.0
FFF: DEBUG: 2017-08-03 01:47:37,614: calibre_plugins.fanficfare_plugin.fff_plugin(1044): url:https://forums.sufficientvelocity.com/threads/39530/
FFF: INFO: 2017-08-03 01:47:37,785: calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_xenforoforum_adapter(239): url: https://forums.sufficientvelocity.com/threads/39530/
FFF: DEBUG: 2017-08-03 01:47:37,785: calibre_plugins.fanficfare_plugin.fanficfare.configurable(1020): #####################################
pagecache(GET) MISS: https://forums.sufficientvelocity.com/threads/39530/
FFF: INFO: 2017-08-03 01:47:38,586: calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_xenforoforum_adapter(244): use useurl: https://forums.sufficientvelocity.com/threads/meddle-not.39530/
FFF: ERROR: 2017-08-03 01:47:40,588: calibre_plugins.fanficfare_plugin.dialogs(680): Exception: {u'comment': u"'href'", u'publisher': None, u'begin': None, u'good': False, u'end': None, u'pubdate': None, u'showerror': True, u'author': [u'Unknown'], u'url': u'https://forums.sufficientvelocity.com/threads/39530/', u'series': u'', u'author_sort': [u'Unknown'], u'site': 'forums.sufficientvelocity.com', u'comments': u'', u'calibre_id': None, u'added': False, u'listorder': 0, u'uniqueurl': u'https://forums.sufficientvelocity.com/threads/39530/[None-None]', u'title': u'Unknown'}:'href'
Traceback (most recent call last):
  File "calibre_plugins.fanficfare_plugin.dialogs", line 668, in do_loop
  File "calibre_plugins.fanficfare_plugin.fff_plugin", line 1134, in prep_download_loop
  File "calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_adapter", line 218, in getStoryMetadataOnly
  File "calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_adapter", line 302, in doExtractChapterUrlsAndMetadata
  File "calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_xenforoforum_adapter", line 275, in extractChapterUrlsAndMetadata
  File "calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_xenforoforum_adapter", line 219, in extract_threadmarks
  File "/home/user/.config/calibre/plugins/FanFicFare.zip/bs4/element.py", line 997, in __getitem__
    return self.attrs[key]
KeyError: 'href'
FFF: DEBUG: 2017-08-03 01:47:55,247: calibre_plugins.fanficfare_plugin.fff_plugin(954): FanFicFare v2.15.0
FFF: DEBUG: 2017-08-03 01:47:55,361: calibre_plugins.fanficfare_plugin.fff_plugin(1044): url:https://forums.spacebattles.com/threads/480163/
FFF: INFO: 2017-08-03 01:47:55,615: calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_xenforoforum_adapter(239): url: https://forums.spacebattles.com/threads/480163/
FFF: DEBUG: 2017-08-03 01:47:55,615: calibre_plugins.fanficfare_plugin.fanficfare.configurable(1020): #####################################
pagecache(GET) MISS: https://forums.spacebattles.com/threads/480163/
FFF: INFO: 2017-08-03 01:47:56,384: calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_xenforoforum_adapter(244): use useurl: https://forums.spacebattles.com/threads/destiny-gospel-of-a-new-genesis-evangelion-destiny.480163/
FFF: ERROR: 2017-08-03 01:47:58,932: calibre_plugins.fanficfare_plugin.dialogs(680): Exception: {u'comment': u"'href'", u'publisher': None, u'begin': None, u'good': False, u'end': None, u'pubdate': None, u'showerror': True, u'author': [u'Unknown'], u'url': u'https://forums.spacebattles.com/threads/480163/', u'series': u'', u'author_sort': [u'Unknown'], u'site': 'forums.spacebattles.com', u'comments': u'', u'calibre_id': None, u'added': False, u'listorder': 0, u'uniqueurl': u'https://forums.spacebattles.com/threads/480163/[None-None]', u'title': u'Unknown'}:'href'
Traceback (most recent call last):
  File "calibre_plugins.fanficfare_plugin.dialogs", line 668, in do_loop
  File "calibre_plugins.fanficfare_plugin.fff_plugin", line 1134, in prep_download_loop
  File "calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_adapter", line 218, in getStoryMetadataOnly
  File "calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_adapter", line 302, in doExtractChapterUrlsAndMetadata
  File "calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_xenforoforum_adapter", line 275, in extractChapterUrlsAndMetadata
  File "calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_xenforoforum_adapter", line 219, in extract_threadmarks
  File "/home/user/.config/calibre/plugins/FanFicFare.zip/bs4/element.py", line 997, in __getitem__
    return self.attrs[key]
KeyError: 'href'
```

**Comment by Zarohk** [comment]
I'm having the same error. FFN is also unaffected.

**Comment by tr13dm03** [comment]
Some more debugging done with the cli.py in dev-mode. With the specific line starting failure.

``` log
FFF: DEBUG: 2017-08-03 18:27:22,894: adapter_gravitytalescom.py(38): No version of feedparser module available, falling back to naive published and updated date
Traceback (most recent call last):
  File "fanficfare/cli.py", line 499, in <module>
    main()
  File "fanficfare/cli.py", line 259, in main
    passed_personalini)
  File "fanficfare/cli.py", line 355, in do_download
    adapter.getStoryMetadataOnly()
  File "/tmp/test-calibre/FanFicFare-2.15.0/fanficfare/adapters/base_adapter.py", line 218, in getStoryMetadataOnly
    self.doExtractChapterUrlsAndMetadata(get_cover=get_cover)
  File "/tmp/test-calibre/FanFicFare-2.15.0/fanficfare/adapters/base_adapter.py", line 302, in doExtractChapterUrlsAndMetadata
    return self.extractChapterUrlsAndMetadata()
  File "/tmp/test-calibre/FanFicFare-2.15.0/fanficfare/adapters/base_xenforoforum_adapter.py", line 275, in extractChapterUrlsAndMetadata
    threadmarks = self.extract_threadmarks(souptag)
  File "/tmp/test-calibre/FanFicFare-2.15.0/fanficfare/adapters/base_xenforoforum_adapter.py", line 219, in extract_threadmarks
    threadmark_rss_dom = parseString(self._fetchUrl(self.getURLPrefix()+'/'+threadmarksa['href'].replace('threadmarks','threadmarks.rss')).encode('utf-8'))
  File "build/bdist.linux-x86_64/egg/bs4/element.py", line 1011, in __getitem__
KeyError: 'href'
```

**Comment by JimmXinu** [comment]
I am away from network, so I will not be able to look at problems for a few
days.

**Comment by sarf** [comment]
I did a brute-force patch-job (not a Python coder by trade, so it is most likely more ugly than it needs to be):
[base_xenforoforum_adapter.py.txt](https://github.com/JimmXinu/FanFicFare/files/1200368/base_xenforoforum_adapter.py.txt)

Apparently they added a threadmarksTrigger link tag without an actual href attribute (probably to handle css-popup menu look and feel or whatnot) which causes problems, but the current code also fails to find any thread mark at all.

**Comment by tr13dm03** [comment]
On Fri, 04 Aug 2017 12:46:05 +0000 (UTC)
sarf <notifications@github.com> wrote:

> I did a brute-force patch-job (not a Python coder by trade, so it is
> most likely more ugly than it needs to be):
> [base_xenforoforum_adapter.py.txt](https://github.com/JimmXinu/FanFicFare/files/1200368/base_xenforoforum_adapter.py.txt)
> 
> Apparently they added a threadmarksTrigger link tag without an actual
> href attribute (probably to handle css-popup menu look and feel or
> whatnot) which causes problems, but the current code also fails to
> find any thread mark at all.
> 

It seems to work. Until someone feels like making a more optimal
version it'll do the trick. I think you should submit a pull request.

**Comment by Zarohk** [comment]
Thank you for the solution. What exactly do I do with this? I'm on a Mac. Do I copy it into terminal, or python, or what?

**Comment by tr13dm03** [comment]
@Zarohk You'd normally generate a patch and apply it to the appropriate file, or just copy the file in its place under the 'adapters' directory (keeps less history of changes, no good). 

Then you'd rebuild the program from source, with the new modifications. Problem is that the current solution isn't perfect and while it allows for downloads, chapter-names are non-functional.

**Comment by Zarohk** [comment]
Um, I really do not know how to code. Any directions you can give me on what to do, or sources to point me at, or should I just wait for the official update?

**Comment by tr13dm03** [comment]
On Sat, 05 Aug 2017 00:34:55 +0000 (UTC)
Zarohk <notifications@github.com> wrote:

> Um, I really do not know how to code. Any directions you can give me
> on what to do, or sources to point me at, or should I just wait for
> the official update?
> 

@Zarokh I would recommend waiting until the official update. The
current one is a partial fix.

**Comment by Zarohk** [comment]
Okay. Thanks for the help, I wish I knew more.

**Comment by JimmXinu** [comment]
Fixed in f39c9d49f8eb179e82b76f776981fe081b6f7537  Test version posted soon.
